Enhance Access Security with METHOD and AUTHORITY support, and add YAML parsing

george-mcintyre opened the pull request:
note: Part of EPICS Secure PVAccess work

Primary changes:
1. **Support for METHOD and AUTHORITY:**
   - Added new fields `METHOD("type"[,...])` and `AUTHORITY("common_name"[,...])` to access security rules for PV authorization based on connection method and certificate authority.
   - Introduced new API functions `asAddClientX`, `asChangeClientX`, and `asTrapWriteBeforeWithDataX` in the `libcom` library.
   - Added a new access type `RPC` to control PV access via RPC (currently operates as `WRITE`).
   - Updated ASG (Access Security Group) processing to incorporate these new fields.

Additional update:
2. **YAML Support:**
   - Added lexer and parser rules for handling YAML-formatted access security configurations.
   - Introduced new functions to manage YAML parsing in ACF files.
   - Introduce new YAML schema file for validation of new YAML format in supported editors

These changes provide fine-grained control over access security based on connection methods and certificate authorities, while also introducing the flexibility of YAML for configuring access security settings.

george-mcintyre committed:
Enhance Access Security with METHOD and AUTHORITY support, and add YAML parsing

Primary changes:
1. **Support for METHOD and AUTHORITY:**
   - Added new fields `METHOD("type"[,...])` and `AUTHORITY("common_name"[,...])` to access security rules for PV authorization based on connection method and certificate authority.
   - Introduced new API functions `asAddClientX`, `asChangeClientX`, and `asTrapWriteBeforeWithDataX` in the `libcom` library.
   - Added a new access type `RPC` to control PV access via RPC (currently operates as `WRITE`).
   - Updated ASG (Access Security Group) processing to incorporate these new fields.

Additional update:
2. **YAML Support:**
   - Added lexer and parser rules for handling YAML-formatted access security configurations.
   - Introduced new functions to manage YAML parsing in ACF files.

These changes provide fine-grained control over access security based on connection methods and certificate authorities, while also introducing the flexibility of YAML for configuring access security settings.

george-mcintyre committed:
Update aslibtest.c with consistent access point names

Modified the INPA and INPB link names in the test to maintain consistency with access point naming conventions. Additionally, updated the commented out calculation rule for better clarity on its intended condition check.

george-mcintyre committed:
Add YAML schema for EPICS Access Security Configuration

Introduce a new YAML schema file for validating EPICS Access Security Configuration. Additionally, update lexer to ignore the schema declaration line during parsing.

george-mcintyre committed:
Handle string end in asLib_lex.l parsing functions

Check for string end before advancing pointers in asSkipToValue and asDupQuotedString. This prevents errors when the expected characters ':' or '"' are not found in the string.

george-mcintyre committed:
Add support for authorization based on whether connection is a TLS connection or not.

Enhanced the access security mechanism to include TLS verification. Updated function signatures and relevant modules to handle TLS status. Added new configuration options and updated documentation with examples for specifying TLS requirements.

anjohnson commented on modules/libcom/test/aslibtest.c:
```diff
@@ -111,6 +157,335 @@ static void testHostNames(void)
+
+    char *buf = NULL;
+    size_t size = 0;
+    FILE *fp = open_memstream(&buf, &size);
```
`open_memstream()` doesn't appear to be available on Windows, which is one of our supported platforms. We might consider dropping some of the older Windows compiler builds before this gets merged, but recent MSVC compilers must be able to build and run the tests.

mdavidsaver replied:
While I appreciate the avoidance of clutter, our tests are meant to run in an `O.<arch>/` sub-directory.  So it would be reasonable, and simpler, to write and read back a file.

mdavidsaver commented on documentation/RELEASE_NOTES.md:
```diff
@@ -38,6 +38,155 @@ For example this will remove the record named "unwanted":
+### Access Security: Certificate-based Authorization
+
+Added a new option and two new fields to `ASG(){RULE(...,<new_option>) {<new_fields>}}`, and added a new access type, `RPC`, to ASG(){RULE(<new_type>)}.  Three new API functions were also added to the `libcom` library to support these new features.  The ability to provide ACF files in EPICS YAML format has been added.
+
```
imo. this is way too long for a release note.  It should be moved to the place(s) where the ACF syntax is described.

mdavidsaver commented on documentation/RELEASE_NOTES.md:
```diff
@@ -38,6 +38,155 @@ For example this will remove the record named "unwanted":
+LIBCOM_API long epicsStdCall asChangeClientX(ASCLIENTPVT asClientPvt,int asl,const char *user,char *method,char *authority,char *host, int isTLS);
+LIBCOM_API void * epicsStdCall asTrapWriteBeforeWithDataX(
+    const char *userid, const char *method, const char *authority, const char *hostid, int isTLS, struct dbChannel *addr,
+    int dbrType, int no_elements, void *data);
```
Remove this verbatim block.  It is redundant to the contents of the header file.

mdavidsaver commented on documentation/RELEASE_NOTES.md:
```diff
@@ -38,6 +38,155 @@ For example this will remove the record named "unwanted":
+New API functions have been added to preserve backwards compatibility with existing clients, and because neither function overloading nor default parameters are supported in the version of C++ used for this release.
+
+#### EPICS YAML format for ACF files
+ACF Files can be specified in YAML format as of this release.  Any ACF file that contains a line that starts with `# EPICS YAML` will introduce yaml processing from that point forwards.  The yaml must include a version number entry.  For the moment only version 1.0 EPICS YAML is supported. For example:
```
I think the YAML part should be a separate PR.  Honestly, I am not enthusiastic about adding a limited parser.

mdavidsaver commented on modules/ca/src/client/caProto.h:
```diff
@@ -146,6 +146,7 @@ typedef ca_uint32_t     caResId;
  */
 #define CA_PROTO_ACCESS_RIGHT_READ  (1u<<0u)
 #define CA_PROTO_ACCESS_RIGHT_WRITE (1u<<1u)
+#define CA_PROTO_ACCESS_RIGHT_CALL (1u<<2u)
```
CA does not, and never will, have a concept of RPC call.  So I think this definition, and the changes under `rsrv/` should be removed.

mdavidsaver commented on modules/libcom/src/as/asLib.h:
```diff
@@ -42,21 +42,28 @@ typedef void (*ASCLIENTCALLBACK) (ASCLIENTPVT,asClientStatus);
 void asTrapWriteAfter(ASCLIENTPVT asClientPvt);
 */
-#define asTrapWriteWithData(asClientPvt, user, host, addr, type, count, data) \
+#define asTrapWriteWithData(asClientPvt, user, host, isTLS, addr, type, count, data) \
```
The number of arguments for the `asTrapWriteWithData()` macro can't change.

mdavidsaver commented on modules/libcom/src/as/asLib.h:
```diff
@@ -42,21 +42,28 @@ typedef void (*ASCLIENTCALLBACK) (ASCLIENTPVT,asClientStatus);
-    ? asTrapWriteBeforeWithData((user), (host), (addr), (type), (count), (data)) \
+    ? asTrapWriteBeforeWithDataX((user), "ca", NULL, (host), (isTLS), (addr), (type), (count), (data)) \
+    : 0)
+#define asTrapWriteWithDataX(asClientPvt, user, method, authority, host, isTLS, addr, type, count, data) \
+    ((asActive && (asClientPvt)->trapMask) \
+    ? asTrapWriteBeforeWithDataX((user), (method), (authority), (host), (isTLS), (addr), (type), (count), (data)) \
```
Firstly, let's add a `static inline` function instead of another "convenience" macro.  Secondly, maybe switch from passing an increasingly long argument list to a `struct`?

mdavidsaver commented on modules/libcom/src/as/asLib.h:
```diff
@@ -126,6 +140,10 @@ LIBCOM_API void * epicsStdCall asTrapWriteBeforeWithData(
     const char *userid, const char *hostid, struct dbChannel *addr,
     int dbrType, int no_elements, void *data);
 
+LIBCOM_API void * epicsStdCall asTrapWriteBeforeWithDataX(
+    const char *userid, const char *method, const char *authority, const char *hostid, int isTLS, struct dbChannel *addr,
+    int dbrType, int no_elements, void *data);
```
ditto...

mdavidsaver commented on modules/libcom/src/as/asLib.h:
```diff
@@ -201,7 +239,10 @@ typedef struct{
+    ELLLIST         methodList; /*List of ASGMETHOD*/
+    ELLLIST         authList;   /*List of ASGAUTHORITY*/
     int             trapMask;
+    int             isTLS; // -1 not set, 0 false, 1 = true
```
I'm not sure I follow what a "TLS" condition is meant to capture, and how it would be used.  Maybe an example would help?

mdavidsaver commented on modules/libcom/src/osi/os/posix/osdThread.c:
```diff
@@ -31,7 +31,7 @@
 
 #if USE_MEMLOCK
-#include <sys/mman.h> 
+#include <sys/mman.h>
```
Remove this extraneous change.

mdavidsaver reviewed (changes requested):
@anjohnson @ralphlange This PR is meant to be a starting point for a discussion about ACF syntax extensions to facilitate the PVA over TLS work.

george-mcintyre committed:
Create temporary file for test dump

This change replaces the use of a memory stream with a temporary file to capture the output of `asDumpFP`. This ensures compatibility across different platforms, accommodating differences in file handling between Windows and Unix-like systems. The test plan count has been incremented to reflect the additional test.

george-mcintyre commented:
Done ✔️

> On 18 Nov 2024, at 11:49, mdavidsaver ***@***.***> wrote:
> 
> 
> @mdavidsaver commented on this pull request.
> 
> In modules/libcom/test/aslibtest.c <https://github.com/epics-base/epics-base/pull/563#discussion_r1847160564>:
> 
> > +  "\tRULE(0,NONE,NOTRAPWRITE)\n"
> +  "\tRULE(1,RPC,NOTRAPWRITE) {\n"
> +  "\t\tUAG(bar)\n"
> +  "\t\tMETHOD(\"x509\",\"ignored\",\"ignored_too\")\n"
> +  "\t\tAUTHORITY(\"Epics Org CA\",\"ignored\",\"ORNL Org CA\")\n"
> +  "\t}\n"
> +  "}\n";
> +
> +static void testDumpOutput(void)
> +{
> +    testDiag("testDumpOutput()");
> +    testOk1(asInitMem(method_auth_config, NULL)==0);
> +
> +    char *buf = NULL;
> +    size_t size = 0;
> +    FILE *fp = open_memstream(&buf, &size);
> While I appreciate the avoidance of clutter, our tests are meant to run in an O.<arch>/ sub-directory. So it would be reasonable, and simpler, to write and read back a file.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/epics-base/epics-base/pull/563#discussion_r1847160564>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABJQ2YV2WHUI7XJKSGGGJE32BJAEZAVCNFSM6AAAAABRSKM46CVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDINBTGU2TMMRTGY>.
> You are receiving this because you were mentioned.
>

george-mcintyre committed:
Remove support for protocol access right RPC

Eliminates the `rsrvCheckCall` function and related code, simplifying access rights management. Also updates relevant documentation and header files to reflect the removal.

george-mcintyre committed:
Remove whitespace change

mdavidsaver commented on modules/libcom/src/as/asTrapWrite.h:
```diff
@@ -33,7 +33,10 @@ struct dbChannel;
+    const char *method; /**< \brief Method of the request (ca, anonymous, or x509. */
+    const char *authority; /**< \brief Authority of the request (common name of the root certificate authority). */
     const char *hostid; /**< \brief Hostid of whoever originated the request. */
+    int isTLS; /**< \brief the connection is TLS encapsulated. */
```
Core Group: Please append new members instead of inserting.

Also, we would like to confirm that caputlog is unaffected.

mdavidsaver commented on documentation/RELEASE_NOTES.md:
```diff
@@ -38,6 +38,145 @@ For example this will remove the record named "unwanted":
+        METHOD("x509")
+        AUTHORITY("SLAC.STANFORD.EDU CA", "STANFORD.EDU CA")
+    }
+}
```
Core Group:

```
UAG(people) { ... }
ASG(secure) {
    RULE(1, WRITE) {
        UAG(people)
        METHOD("x509")
        AUTHORITY("SLAC.STANFORD.EDU CA", "STANFORD.EDU CA")
    }
}
```

Should `METHOD` and `AUTHORITY` be part of `UAG()`?

What about `HAG`?

Commit:
Add Flux v2.4.0 component manifests

Commit:
Add Flux sync manifests

mdavidsaver commented:
I am switching this to Draft status pending a design discussion and updates.

george-mcintyre committed:
Add GitHub Actions workflow for multi-platform Docker builds

This workflow automates building and pushing EPICS base Docker images for multiple platforms (amd64, arm64). It uses repositories and configurations for both EPICS base and PVXS, triggered by manual dispatch or workflow calls.

george-mcintyre committed:
Add GitHub Actions workflow for multi-platform Docker builds

This workflow automates building and pushing EPICS base Docker images for multiple platforms (amd64, arm64). It uses repositories and configurations for both EPICS base and PVXS, triggered by manual dispatch or workflow calls.

george-mcintyre committed:
Add GitHub Actions workflow for multi-platform Docker builds

This workflow automates building and pushing EPICS base Docker images for multiple platforms (amd64, arm64). It uses repositories and configurations for both EPICS base and PVXS, triggered by manual dispatch or workflow calls.

george-mcintyre committed:
Add GitHub Actions workflow for multi-platform Docker builds

This workflow automates building and pushing EPICS base Docker images for multiple platforms (amd64, arm64). It uses repositories and configurations for both EPICS base and PVXS, triggered by manual dispatch or workflow calls.

george-mcintyre committed:
Add GitHub Actions workflow for multi-platform Docker builds

This workflow automates building and pushing EPICS base Docker images for multiple platforms (amd64, arm64). It uses repositories and configurations for both EPICS base and PVXS, triggered by manual dispatch or workflow calls.

george-mcintyre committed:
Add GitHub Actions workflow for multi-platform Docker builds

This workflow automates building and pushing EPICS base Docker images for multiple platforms (amd64, arm64). It uses repositories and configurations for both EPICS base and PVXS, triggered by manual dispatch or workflow calls.

george-mcintyre committed:
Add GitHub Actions workflow for multi-platform Docker builds

This workflow automates building and pushing EPICS base Docker images for multiple platforms (amd64, arm64). It uses repositories and configurations for both EPICS base and PVXS, triggered by manual dispatch or workflow calls.

george-mcintyre committed:
Add GitHub Actions workflow for multi-platform Docker builds

This workflow automates building and pushing EPICS base Docker images for multiple platforms (amd64, arm64). It uses repositories and configurations for both EPICS base and PVXS, triggered by manual dispatch or workflow calls.

george-mcintyre commented:
Replaced by #618, split into two PRs #618 deals only with METHOD and AUTHORITY feature addition
